geth/node/peerpool.go

```diff
@@ -0,0 +1,247 @@
+func (p *PeerPool) Start(server *p2p.Server) error {
+	if server.DiscV5 == nil {
+		return ErrDiscv5NotRunning
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	// 2 goroutines per each topic
+	p.wg.Add(len(p.config) * 2)
+	p.quit = make(chan struct{})
+	// sync periods are stored because we need to close them once pool is stopped
+	p.init()
+
+	for topic := range p.config {
+		topic := topic
+		period := make(chan time.Duration, 2)
+		p.syncPeriods = append(p.syncPeriods, period)
+		found := make(chan *discv5.Node, 10)
+		lookup := make(chan bool, 10)
+		events := make(chan *p2p.PeerEvent, 20)
+		subscription := server.SubscribeEvents(events)
+		p.subscriptions = append(p.subscriptions, subscription)
+		go func() {
+			server.DiscV5.SearchTopic(topic, period, found, lookup)
+			p.wg.Done()
+		}()
+
+		go func() {
+			p.handlePeersFromTopic(server, topic, period, found, lookup, events)
+			p.wg.Done()
+		}()
+	}
+	return nil
+}
+
+func (p *PeerPool) handlePeersFromTopic(server *p2p.Server, topic discv5.Topic, period chan time.Duration, found chan *discv5.Node, lookup chan bool, events chan *p2p.PeerEvent) {
+	limits := p.config[topic]
+	fast := true
+	period <- p.fastSync
+	connected := 0
+	for {
+		select {
+		case <-p.quit:
+			return
+		case node := <-found:
+			if p.processFoundNode(server, connected, topic, node) {
+				connected++
+			}
+			// switch period only once
+			if fast && connected >= limits[0] {
+				period <- p.slowSync
+				fast = false
+			}
+		case <-lookup:
+			// just drain this channel for now, it can be used to intellegently
+			// limit number of kademlia lookups
+		case event := <-events:
+			if event.Type == p2p.PeerEventTypeDrop {
+				if !p.processDisconnectedNode(server, topic, event.Peer) {
+					connected--
+				}
+				// switch period only once
+				if !fast && connected < limits[0] {
+					period <- p.fastSync
+					fast = true
+				}
+			}
+
+		}
+	}
+}
+
+// processFoundNode called when node is discovered by kademlia search query
+// 2 important conditions
+// 1. every time when node is processed we need to update discoveredTime and reset dropped boolean.
+//    peer will be considered as valid later only if it was discovered < 90s ago and wasn't dropped recently
+// 2. if peer is connected or if max limit is reached we are not a adding peer to p2p server
+func (p *PeerPool) processFoundNode(server *p2p.Server, currentlyConnected int, topic discv5.Topic, node *discv5.Node) (connected bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	peersTable := p.peers[topic]
+	limits := p.config[topic]
+	if info, exist := peersTable[node.ID]; exist {
+		info.discoveredTime = mclock.Now()
+		info.dropped = false
+	} else {
+		peersTable[node.ID] = &peerInfo{
+			connected:      false,
+			dropped:        false,
+			discoveredTime: mclock.Now(),
+			node: discover.NewNode(
+				discover.NodeID(node.ID),
+				node.IP, node.UDP, node.TCP),
+		}
+	}
+	if currentlyConnected < limits[1] && !peersTable[node.ID].connected {
```

Comment:
Shouldn't it be `<= limits[1]`? If `limits[1]` was 1, then no peer would end up being added, right?

Reply:
currentlyConnected could be 0 and if max limit is 1, then peer will be connected. if current is 1 and max limit is 1... seems to be correct, no?

Reply:
you're right, nvm